client/connect.go

```diff
@@ -95,13 +98,17 @@ func (client *Client) StartWatchdog() {
 	startTime := time.Now()
 	for {
 		// it didn't sent anything before
-		if (client.lastSent == time.Time{}) {
-			if startTime.Add(10 * time.Minute).Before(time.Now()) {
+		client.blockedM.Lock()
+		{
+			if (client.lastSent == time.Time{}) {
+				if startTime.Add(10 * time.Minute).Before(time.Now()) {
+					break
+				}
+			} else if client.lastSent.Add(10 * time.Minute).Before(time.Now()) {
 				break
 			}
```

Comment:
This here is another locked read. I think you can remove the lock and/or extract `lastSent` in a variable so it's read only once per iteration.

Reply:
there're writes below in line 107  `client.lastSent.Add(10 * time.Minute).Before(time.Now())`